### PR TITLE
chore: Benchmark WAL formats

### PR DESCRIPTION
**Pull request by pauldix**
This PR started as a result of my desire to move the WAL Buffer over to using JSON as the format. My reasoning was as follows:
* The Flatbuffers generated code is painful to work with in Rust.
* The WAL Segment files in Object Storage are part of the public facing IOx API, thus having an easier to work with format would be something that users would appreciate.
* This [writeup indicates that JSON is faster than Flatbuffers in their tests](https://blog.logrocket.com/rust-serialization-whats-ready-for-production-today/), so maybe it's not really worth it?
* Whatever the replication format is, it's going to get converted into different Rust structs to make it useful for querying or writing when it gets converted to Parquet for storage in object store. My thought was that this would dominate any deserialization time.

Mainly, I was wondering if Flatbuffers was a premature optimization that we didn't really need. Maybe we could simplify the code base and the WAL Segment format all in one go?

My conclusion after running through this exercise is that we should keep the Flatbuffers format and, if needed, build more helper functions around it. We might also want to build C wrappers that can be used in other languages to make it easier to work with natively.

## Detailed Description
This PR is split into two commits. The first commit (7586fc3968d4231b28416f8bb1e9dbc44368ad9e) adds some benchmarking for the existing Flatbuffers structure. We'll likely want to keep this around once people have a chance to review this PR and comment on it. If so, I can cherry-pick it into another PR to get it into shape for merging into main. The second commit (ba48064a7513a1cd69e6be93d78fd29b4bd4a7bf) adds some methods in the benchmark that mirror the existing ReplicatedWrite structure, but with JSON so that we can compare its performance. In this commit I commented out the checksum computation that the Flatbuffers implementation does so the comparison is fair.

I wanted to separate the benchmarks into discrete things that we'll do in a production system. Sometimes we'd have these end-to-end, but many times a server will only be doing one of these steps. The benchmarks test:
* `lines_to_replicated_write` - taking already parsed line protocol and creating a ReplicatedWrite struct with it.
* `replicated_write_into_bytes` - convert a ReplicatedWrite into bytes so that it can be sent over the network or into a segment file.
* `bytes_into_struct` - deserialize a replicated write and load it into a toy DB struct. The idea here is that a ReplicatedWrite will always have to be converted over to either the MutableBuffer, ReadBuffer, or RecordBatch. None of the underlying crates will use the ReplicatedWrite as is. I implemented something basic that uses a dictionary per partition to avoid a bunch of String allocations taking up the lion's share of the time.

The easiest way to see the results is to pull down this branch and do from the project root:

```
cd data_types
cargo bench
open target/criterion/report/index.html 
```

## Results Summary
JSON is slightly faster in the lines to replicated write benchmark at 4.9% to 8.4% faster. This is about 250 microseconds in a bench that takes 2.75 to 2.97 milliseconds. The impact is slightly more pronounced in scenarios with a single partition and either many tables or tags (which would likely be most commmon). But it's not much.

In the bytes to struct benchmark, JSON time is taken up largely by deserializing the bytes, giving Flatbuffers over a 3x advantage.

The results of replicated write to bytes are predictable in that Flatbuffers takes no time (< 1ns) while JSON serialization takes around 820 microseconds for the same thing. While 820 microseconds seems small, it will likely add up with larger segment files (i.e. not microbenchmarks) and particularly for replication or subscriptions that take the replicated writes as is (i.e. they contain no filter).

## Closing Thoughts
There are a few other advantages that Flatbuffers has over JSON. I think a declarative schema is a win. We should take advantage of this and version them so we can evolve over time without breaking users. Also, we plan to support blob types in IOx, which will be trivial in Flatbuffers, while in JSON we'd end up doing something ugly like base64 encoding. Generally, data types in Flatbuffers are more specific. We won't have a problem representing `u64` or the range of `float` values for instance.

One thing this doesn't reflect is that subscriptions will likely filter out portions of a ReplicatedWrite. This means that it will need to be processed so that a new one can be built and then serialized. So we'll be creating new data as we go. This would likely remove a bit of the performance advantage of Flatbuffers over JSON, but I think it'll still be much faster than we'd otherwise get with JSON.

I didn't get very far into this effort before I started reaching the conclusion that Flatbuffers would be the better choice. It seems obvious in retrospect, but I thought that other factors would likely drown out any performance differences between JSON and Flatbuffers (or make the difference small enough that ease of use would win out). But I decided to forge ahead anyway so that we'd have this documented and to convince myself.

Unless I'm missing something?

### CODE REVIEW

File: data_types/src/data.rs [hunk 1]
```diff
@@ -148,9 +172,10 @@ pub fn lines_to_replicated_write(
         lines,
     );
 
-    let mut hasher = Hasher::new();
-    hasher.update(&entry_bytes);
-    let checksum = hasher.finalize();
+    // let mut hasher = Hasher::new();
+    // hasher.update(&entry_bytes);
+    // let checksum = hasher.finalize();
+    let checksum = 0u32;
```
Comment (alamb) [inline data_types/src/data.rs:178]: why did you avoid calculating the checksum? In particular, I wonder if the checksum calculation itself took up a significant portion of the time?

File: data_types/benches/benchmark.rs [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+}
+
+fn replicated_write_into_bytes(c: &mut Criterion) {
+    run_group(
+        "replicated_write_into_bytes",
+        c,
+        |lines, rules, config, base_name, group| {
+            let id = BenchmarkId::new(base_name, "fb");
+            group.bench_with_input(id, &config, |b, config| {
+                let write = lines_to_rw(0, 0, lines, rules);
```
Comment (alamb) [inline data_types/benches/benchmark.rs:83]: The description of this PR lead me to believe that the "replicated_writes_into_bytes" benchmark would only be testing the time taken to convert the RepliatedWrites into json -- this benchmark also appears to include the time required to turn the parsed_lines into ReplicatedWrites in the first place.

**Review by alamb — COMMENTED** [review]
> One thing this doesn't reflect is that subscriptions will likely filter out portions of a ReplicatedWrite. This means that it will need to be processed so that a new one can be built and then serialized. So we'll be creating new data as we go. This would likely remove a bit of the performance advantage of Flatbuffers over JSON, but I think it'll still be much faster than we'd otherwise get with JSON.

One thought I had, which may or may not be useful, would be rather than copy a portion of a ReplicatedWrite, we could send the original along with a separate list of "which writes to take (or delete)" -- that would save processing time at the expense of sending bytes that might get ignored by the receiver. Whether this is a good idea is probably a function of the number of writes that got filtered out

All in all, this is a nice result and writeup -- thanks @pauldix

File: data_types/src/data.rs (hunk 1 above)
Comment (pauldix) [reply data_types/src/data.rs:178]: It made a difference, but it wasn't very large. There's actually no need to checksum an individual replicated write as they get sent in individual requests or in bulk in a segment that gets shipped to object store. Some of this is cruft from how it evolved over time with the previous disk based WAL. I'll likely refactor this soon to reflect the new design of segments in object store.

File: data_types/benches/benchmark.rs (hunk 2 above)
Comment (pauldix) [reply data_types/benches/benchmark.rs:83]: That's right. I wanted to see if it took longer to construct the Flatbuffers payload than it took to construct plain old Rust structs. Answer: yes, but not by much.

**Comment by pauldix** [comment]
@alamb that's a good point about subscriptions. And given that the other side (the subscriber) won't spend cycles deserializing, that makes it much more compelling. If you're in the same AZ, network is cheap and fast.

File: data_types/benches/benchmark.rs [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+    }
+
+    fn measurement_count(&self) -> usize {
+        let mut measurements = BTreeMap::new();
+
+        for p in self.partitions.values() {
+            for t in p.tables.values() {
+                measurements.insert(t.name.to_string(), ());
+            }
+        }
```
Comment (carols10cents) [inline data_types/benches/benchmark.rs:533]: This is pretty much what `BTreeSet` does; I'd recommend using that:

```suggestion
        let mut measurements = BTreeSet::new();

        for p in self.partitions.values() {
            for t in p.tables.values() {
                measurements.insert(t.name.to_string());
            }
        }
```

**Review by carols10cents — COMMENTED** [review]
Just some small thoughts!

File: data_types/benches/benchmark.rs (hunk 3 above)
Comment (pauldix) [reply data_types/benches/benchmark.rs:533]: Added in 5769558cc38be26d503975020203f059b7376ef8 in #607 as this PR will be getting closed out.

File: data_types/benches/benchmark.rs [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+    }
+}
+
+impl fmt::Display for Config {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "lines: {} ({} per partition) | partitions: {} | tables per: {} | unique tag values per: {}",
+            self.line_count,
+            self.line_count / self.partition_count as i64,
```
[the remark this answers is not in the document]
Comment (pauldix) [reply data_types/benches/benchmark.rs:44]: Indeed, in 5769558cc38be26d503975020203f059b7376ef8 in #607 which will replace this PR.

File: data_types/benches/benchmark.rs [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+use std::fmt;
+use std::time::Duration;
+
+const NEXT_ENTRY_NS: i64 = 1_000_000_000;
+const STARTING_TIMESTAMP_NS: i64 = 0;
+
+#[derive(Debug)]
+struct Config {
+    // total number of rows written in, regardless of the number of partitions
+    line_count: i64,
```
[the remark this answers is not in the document]
Comment (pauldix) [reply data_types/benches/benchmark.rs:23]: Ah, I think just some cruft from various iterations on this. Fixed in 5769558cc38be26d503975020203f059b7376ef8 in #607 which will replace this PR.

**Comment by pauldix** [comment]
Pulled the usable bench code into #607. Closing this out, but happy to continue to discuss if there are any other comments or thoughts.

**Pull request closed without merging**